mne/io/edf/edf.py

```diff
@@ -296,114 +320,47 @@ def _parse_tal_channel(tal_channel_data):
-        edf_info['digital_min'] = digital_min
-        digital_max = np.array([float(fid.read(8).decode())
-                                for ch in channels])[include]
-        prefiltering = [fid.read(80).strip().decode() for ch in channels][:-1]
-        highpass = np.ravel([re.findall('HP:\s+(\w+)', filt)
-                             for filt in prefiltering])
-        lowpass = np.ravel([re.findall('LP:\s+(\w+)', filt)
-                            for filt in prefiltering])
-
-        # number of samples per record
-        n_samps = np.array([int(fid.read(8).decode()) for ch
-                            in channels])
-        edf_info['n_samps'] = n_samps
-
-        fid.read(32 * nchan).decode()  # reserved
-        assert fid.tell() == header_nbytes
-
-        fid.seek(0, 2)
-        n_bytes = fid.tell()
-        n_data_bytes = n_bytes - header_nbytes
-        total_samps = (n_data_bytes // 3 if subtype == '24BIT'
-                       else n_data_bytes // 2)
-        read_records = total_samps // np.sum(n_samps)
-        if n_records != read_records:
-            warn('Number of records from the header does not match the file '
-                 'size (perhaps the recording was not stopped before exiting).'
-                 ' Inferring from the file size.')
-            edf_info['n_records'] = n_records = read_records
-
-    n_samps = n_samps[include]
-    physical_ranges = physical_max - physical_min
-    cals = digital_max - digital_min
-
-    if edf_info['subtype'] in ('24BIT', 'bdf'):
-        edf_info['data_size'] = 3  # 24-bit (3 byte) integers
+    # Read header from file
+    ext = os.path.splitext(fname)[1][1:].lower()
+    logger.info('%s file detected' % ext.upper())
+    if ext in ('bdf', 'edf'):
+        edf_info = _read_edf_header(fname, stim_channel, annot, annotmap,
+                                    exclude, preload)
+    elif ext in ('gdf'):
+        if annot is not None:
+            warn('Annotations not yet supported for GDF files.')
+        edf_info = _read_gdf_header(fname, stim_channel, exclude)
+        edf_info['annot'] = None
+        edf_info['annotmap'] = None
     else:
-        edf_info['data_size'] = 2  # 16-bit (2 byte) integers
-
-    # Creates a list of dicts of eeg channels for raw.info
-    logger.info('Setting channel info structure...')
-    chs = list()
-
+        raise NotImplementedError(
+            'Only GDF, EDF, and BDF files are supported, got %s.' % ext)
+
+    include = edf_info['include']
+    ch_names = edf_info['ch_names']
+    n_samps = edf_info['n_samps'][include]
+    nchan = edf_info['nchan']
+    physical_ranges = edf_info['physical_max'] - edf_info['physical_min']
+    cals = edf_info['digital_max'] - edf_info['digital_min']
+    if np.any(~np.isfinite(cals)):
+        idx = np.where(~np.isfinite(cals))[0]
+        warn('Scaling factor is not defined in following channels:\n' +
+             str(idx).strip('[]'))
```

Comment:
This is actually unsafe if one has `[` in a channel name (rare I know but...) you can do something more robust and with a nicer output as `', '.join(ch_names[ii] for ii in idx)`.